src/chase/impl/basic.rs

```diff
@@ -221,7 +221,8 @@ impl ModelTrait for Model {
                     false
                 } else {
                     let terms: Vec<WitnessTerm> = terms.into_iter().map(|e| e.unwrap().clone().into()).collect();
-                    self.facts.contains(&Observation::Fact { relation: (*relation).clone(), terms })
+                    let obs = Observation::Fact { relation: (*relation).clone(), terms };
```

Comment:
I'm surprised you need the clone, this should just work fine as well:

```suggestion
                    let obs = Observation::Fact { relation: relation.clone(), terms };
```